x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,421 @@
+        }
+    }
+
+    private void handleFailure(Exception e, Runnable task) {
+        assert e != null;
+        if (shouldRetry(e)) {
+            if (isStopped() == false && retryCounter.incrementAndGet() <= RETRY_LIMIT) {
+                LOGGER.warn("error during follow shard task, retrying...", e);
+                scheduler.accept(RETRY_TIMEOUT, task);
+            } else {
+                markAsFailed(new ElasticsearchException("retrying failed [" + retryCounter.get() +
+                    "] times, aborting...", e));
+            }
+        } else {
+            markAsFailed(e);
+        }
+    }
+
+    private boolean shouldRetry(Exception e) {
+        // TODO: What other exceptions should be retried?
+        return NetworkExceptionHelper.isConnectException(e) ||
+            NetworkExceptionHelper.isCloseConnectionException(e) ||
+            e instanceof ActionTransportException ||
+            e instanceof NodeClosedException ||
+            e instanceof UnavailableShardsException ||
+            e instanceof NoShardAvailableActionException;
+    }
+
+    // These methods are protected for testing purposes:
+    protected void updateMapping(LongConsumer handler) {
+        Index leaderIndex = params.getLeaderShardId().getIndex();
+        Index followIndex = params.getFollowShardId().getIndex();
+
+        ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
+        clusterStateRequest.clear();
+        clusterStateRequest.metaData(true);
+        clusterStateRequest.indices(leaderIndex.getName());
+
+        leaderClient.admin().cluster().state(clusterStateRequest, ActionListener.wrap(clusterStateResponse -> {
+            IndexMetaData indexMetaData = clusterStateResponse.getState().metaData().getIndexSafe(leaderIndex);
+            assert indexMetaData.getMappings().size() == 1;
+            MappingMetaData mappingMetaData = indexMetaData.getMappings().iterator().next().value;
+
+            PutMappingRequest putMappingRequest = new PutMappingRequest(followIndex.getName());
+            putMappingRequest.type(mappingMetaData.type());
+            putMappingRequest.source(mappingMetaData.source().string(), XContentType.JSON);
+            followerClient.admin().indices().putMapping(putMappingRequest, ActionListener.wrap(
+                putMappingResponse -> handler.accept(indexMetaData.getVersion()),
+                e -> handleFailure(e, () -> updateMapping(handler))));
+        }, e -> handleFailure(e, () -> updateMapping(handler))));
+    }
+
+    protected void innerSendBulkShardOperationsRequest(Translog.Operation[] operations,
+                                                     LongConsumer handler,
+                                                     Consumer<Exception> errorHandler) {
+        final BulkShardOperationsRequest request = new BulkShardOperationsRequest(params.getFollowShardId(), operations);
+        followerClient.execute(BulkShardOperationsAction.INSTANCE, request,
```

Comment:
you can use `ActionListener.wrap(r -> handler.accept(r.getLocalCheckpoint()), errorHandler::accept)` instead